### DISCUSSION TITLE
CAIP-19: Defining namespaces for fiat currencies

### DISCUSSION THREAD

**Discussion by danroc** [ChainAgnostic/CAIPs#338] [CAIPs]
I’d like to propose reserving some namespaces to allow CAIP-19 to include fiat currencies. This would simplify certain APIs that need to handle both cryptocurrencies and fiat currencies.

**Proposed format:**

```
asset_type:        chain_id + "/" + asset_namespace + ":" + asset_reference
chain_id:          namespace + ":" + reference
namespace:         "fiat"
reference:         country_code
asset_namespace:   "currency"
asset_reference:   currency_code
```

- `country_code`: A two-letter lowercase code as defined by ISO 3166-1 alpha-2, representing the emitter country. The European Union would be an exception and is represented by "eu".

- `asset_namespace`: A three-letter uppercase code as defined by ISO 4217.

**Examples:**

```
# Euro
fiat:eu/currency:eur

# United States Dollar
fiat:us/currency:usd

# Brazilian Real
fiat:br/currency:brl

# Japanese Yen
fiat:jp/currency:jpy
```

**Comment by bumblefudge** [comment]
Hey! I think this is an interesting idea and probably very useful to people, but i'm wondering if there might not be a certain "trust anchor" (or more technically, a specification-anchor) more precise than country-codes as contract addresses. I think there are a lot of corner-cases where currencies don't map 1:1 to countries, or countries have different "APIs" for their issuance systems corresponding to different governance structures for their central banks, making the analogy a little shaky.  If anything, the reason people like blockchains so much is that they are written from the APIs up to the humans, not from the messy humans down! 

Compare [the actual fed](https://www.federalreserve.gov/aboutthefed/files/board-of-governors-org-chart.pdf) to the [smart contract](https://gist.github.com/chappjc/350aafb9031f7a66986967bf8ab67d38) that lives at [one canonical address per EVM chain](https://etherscan.io/address/0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48) governing not the currency itself but one easily-modeled onchain subset of that currency...

My initial thought is to find a similar subset of the `fiat` namespace that is a little more stable and bounded than the messy realworld chaos of actual currencies. Take, for instance, the various realworld [RTGS systems](https://www.swift.com/our-solutions/market-infrastructures/swift-for-hvps-rtgs) built on the SWIFT messaging standard (a "virtual machine" if ever there was one!), which already have [well-documented APIs](https://www.swift.com/myswift/ordering/order-products-services/eba-clearing-services) and SLAs and excludes all the messiest currencies with contested issuance systems or corresponding to unstable constituencies like civil war factions or contested territories. Something like this might make a better namespace than the entire space of fiat per se, interactions with which are much harder to design a good bridge to or oracle for 😄

**Reply by danroc** [reply]
Hey, thanks for the feedback! I completely agree that using country codes introduces instability. My intention in including them was to define a "zone of application" and align with CAIP-2, which requires a `reference`. I proposed ISO 3166 because it’s used for ccTLDs, with some well-known exceptions like `.eu` and `.uk`, which also seem relevant to our problem.

I haven’t used the APIs you mentioned, but from this [spec](https://community.developer-test.swift.com/sites/default/files/2023-10/SWIFT-API-boe_rtgs_transactions_api_internal-1.0.0-swagger.yaml), it appears they also rely on ISO 4217, which is what I suggested for the `asset_reference`. Perhaps `iso4217` could also work as the `asset_namespace`.

Do you have any suggestions for a subset of `fiat`? We could drop the "geographic" aspect of my initial proposal and use something like `fiat:currency/iso4217:eur`, for example. This avoids the need to decide which fiat currencies to support and sidesteps assumptions about geopolitical factors.

Would this approach align with what you had in mind, or do you have another idea for the `fiat:currency` part?

**Reply by bumblefudge** [reply]
sorry for the delay, i got sidetracked getting my hands on the ISO doc itself.  Apparently the registry of valid iso4217 values (and their official long-forms/human-readable names) can be downloaded here:
https://www.six-group.com/en/products-services/financial-information/data-standards.html (scroll down to "current and historical lists") 

my thinking is that we do this as a namespace called "SWIFT" rather than "fiat" because the actual ground-truth rules of "fiat" are kind of fractal and unknowable, whereas SWIFT is a subset of digital fiat that is relatively machine-readable and simpler to model.  ISO 4217 is the "slip44" equivalent, and I don't have a strong preference for recognizing the 3-digit numeric codes, the 3-digit alphanumerics, or both.

so, in summary, i'm thinking
`swift/iso4217:eur` or `swift:0/iso4217:eur` could be the CAIP:19 syntax, i.e. `0` could be, for now, the only valid "chainId" (to signify offchain/outside the SWIFT system), unless/until there are "chainId"s worth defining?

I could start a PR on namespaces.chainagnostic.org containing this if you'd like to have a strawman/starting point...

**Reply by danroc** [reply]
I agree with your point. My preference for `fiat:currency` is mostly because it's the usual term, but `swift:0` is a good option too. I don't have any strong opinions about using `0` as the chain reference to comply with CAIP-19, it makes sense.

> I could start a PR on namespaces.chainagnostic.org containing this if you'd like to have a strawman/starting point...

Yes, please. That would be much appreciated!

**Reply by bumblefudge** [reply]
https://github.com/ChainAgnostic/namespaces/pull/132/files

It's a branch on the main repo so you should be able to open a sub-PR if it's simpler or faster than change requests (as could anyone else who stumbles across it and knows ISO20022 or, ideally, is already using it for on-/off-ramp accounting!)
